src/QsCompiler/SyntaxProcessor/CapabilityInference/TypeAnalyzer.fs

```diff
@@ -0,0 +1,210 @@
+
+    let capability =
+        Seq.map scenarioCapability scenarios |> Seq.fold TargetCapability.merge TargetCapability.bottom
+
+    if capability = TargetCapability.bottom then
+        None
+    else
+        let diagnose (target: Target) =
+            if TargetCapability.subsumes target.Capability capability then
+                None
+            else
+                let unsupported =
+                    Seq.filter (scenarioCapability >> TargetCapability.subsumes target.Capability >> not) scenarios
+
+                let description = Seq.map describeScenario unsupported |> String.concat ", "
+                let args = [ target.Name; description ]
+                let range = QsNullable.defaultValue Range.Zero range
+                QsCompilerDiagnostic.Error(ErrorCode.UnsupportedClassicalCapability, args) range |> Some
+
+        Some
+            {
+                Capability = capability
+                Diagnose = diagnose
+                Properties = ()
+            }
+
+let rec flattenTuple =
+    function
+    | QsTupleItem x -> Seq.singleton x
+    | QsTuple xs -> Seq.collect flattenTuple xs
+
+let paramPatterns context callable =
+    flattenTuple callable.ArgumentTuple
+    |> Seq.choose (fun param ->
+        let relativeRange = TypeRange.tryRange param.Type.Range
+        let range = (callable.Location, relativeRange) ||> QsNullable.Map2(fun l r -> l.Offset + r)
+        createPattern context Param range param.Type)
+
+let returnPattern context callable =
+    let ty = callable.Signature.ReturnType
+    let relativeRange = TypeRange.tryRange ty.Range
+    let range = (callable.Location, relativeRange) ||> QsNullable.Map2(fun l r -> l.Offset + r)
+    createPattern context Return range ty
+
+let analyzer (action: SyntaxTreeTransformation -> _) : _ seq =
+    let transformation = LocatingTransformation TransformationOptions.NoRebuild
+    let context = ref { IsEntryPoint = false; StringLiteralsOk = false }
+    let patterns = ResizeArray()
+
+    transformation.Namespaces <-
+        { new NamespaceTransformation(transformation, TransformationOptions.NoRebuild) with
+            override _.OnCallableDeclaration callable =
+                let isEntryPoint = Seq.exists BuiltIn.MarksEntryPoint callable.Attributes
+                use _ = local { context.Value with IsEntryPoint = isEntryPoint } context
+
+                let callable = base.OnCallableDeclaration callable
+                paramPatterns context.Value callable |> patterns.AddRange
+                returnPattern context.Value callable |> Option.iter patterns.Add
+                callable
+        }
+
+    transformation.Statements <-
+        { new StatementTransformation(transformation, TransformationOptions.NoRebuild) with
+            override _.OnStatement statement =
+                match statement.Statement with
+                | QsFailStatement _ ->
+                    use _ = local { context.Value with StringLiteralsOk = true } context
```

Comment:
So we are saying stings are ok in messages and fail statements, though we are currently forbidding fail statements. I think that will work, and we may want to reenable failstatements.

Reply:
I did this to avoid a double error for `fail` (one for the `fail` statement itself, and one for the message string). Are `fail` statements supported?

Reply:
No, they are definitively not supported on hardware, but it is a question of whether we should allow them at the language level and fail after qat validation if they have not been optimized away (i.e. fail statements cannot be supported by the runtime, but plenty of fails check compile time information). Writing this down, I think I am coming to the conclusion that we should not allow them, and instead think about where to improve the language/libraries such that these fail statements are not needed.